### ISSUE TITLE
bug: griffe 0.39.0 - cannot collect data encapsulated in two folders with the same name

### ISSUE THREAD

**Issue by IronTopi** [mkdocstrings/mkdocstrings#646]
### Description of the bug

When running `mkdocs` and using `mkdocstrings-python` with `griffe 0.39.0`, a certain folder-structure results in an error.
This used to work with `griffe 0.38.0` and a manual downgrade of griffe makes the whole process run again.

**The problem occurs when a python-module is being loaded that is located in an at least two-folders deep structure where the two folders have the same name:**

- `aaa.bbb.pythonfile`: works fine
- `aaa.aaa.pythonfile`: results in error `Could not collect 'aaa.aaa.pythonfile'`

This issue only occurs in the `mkdocs`-process.
A `griffe dump aaa.aaa.pythonfile` works fine.

**Please feel free to move/redirect to the any of the related repos if this would be better suited!!!**

### To Reproduce

I set up a minimal example the provides a file structure to reproduce that error:

 ```
git clone https://github.com/IronTopi/griffe-bug
cd griffe-bug
python -m venv .venv
. .venv/bin/activate
pip install -r requirements.txt
mkdocs build -v
```

### Full traceback

<details><summary>Full traceback</summary>

```
mkdocs build -v
DEBUG   -  Loading configuration file: /griffe-bug/mkdocs.yml
DEBUG   -  Loaded theme configuration for 'mkdocs' from
           '/griffe-bug/.venv/lib/python3.12/site-packages/mkdocs/themes/mkdocs/mkdocs_theme.yml':
           {'static_templates': ['404.html'], 'locale': 'en',
           'include_search_page': False, 'search_index_only': False,
           'highlightjs': True, 'hljs_languages': [], 'hljs_style':
           'github', 'navigation_depth': 2, 'nav_style': 'primary',
           'analytics': {'gtag': None}, 'shortcuts': {'help': 191,
           'next': 78, 'previous': 80, 'search': 83}}
DEBUG   -  Config value 'config_file_path' = '/griffe-bug/mkdocs.yml'
DEBUG   -  Config value 'site_name' = 'My Docs'
DEBUG   -  Config value 'nav' = None
DEBUG   -  Config value 'pages' = None
DEBUG   -  Config value 'exclude_docs' = None
DEBUG   -  Config value 'not_in_nav' = None
DEBUG   -  Config value 'site_url' = None
DEBUG   -  Config value 'site_description' = None
DEBUG   -  Config value 'site_author' = None
DEBUG   -  Config value 'theme' = Theme(name='mkdocs',
           dirs=['/griffe-bug/.venv/lib/python3.12/site-packages/mkdocs/themes/mkdocs',
           '/griffe-bug/.venv/lib/python3.12/site-packages/mkdocs/templates'],
           static_templates={'sitemap.xml', '404.html'},
           name='mkdocs', locale=Locale(language='en', territory=''),
           include_search_page=False, search_index_only=False,
           highlightjs=True, hljs_languages=[], hljs_style='github',
           navigation_depth=2, nav_style='primary',
           analytics={'gtag': None}, shortcuts={'help': 191, 'next':
           78, 'previous': 80, 'search': 83})
DEBUG   -  Config value 'docs_dir' = '/griffe-bug/docs'
DEBUG   -  Config value 'site_dir' = '/griffe-bug/site'
DEBUG   -  Config value 'copyright' = None
DEBUG   -  Config value 'google_analytics' = None
DEBUG   -  Config value 'dev_addr' =
           _IpAddressValue(host='127.0.0.1', port=8000)
DEBUG   -  Config value 'use_directory_urls' = True
DEBUG   -  Config value 'repo_url' = None
DEBUG   -  Config value 'repo_name' = None
DEBUG   -  Config value 'edit_uri_template' = None
DEBUG   -  Config value 'edit_uri' = None
DEBUG   -  Config value 'extra_css' = []
DEBUG   -  Config value 'extra_javascript' = []
DEBUG   -  Config value 'extra_templates' = []
DEBUG   -  Config value 'markdown_extensions' = ['toc', 'tables',
           'fenced_code']
DEBUG   -  Config value 'mdx_configs' = {}
DEBUG   -  Config value 'strict' = False
DEBUG   -  Config value 'remote_branch' = 'gh-pages'
DEBUG   -  Config value 'remote_name' = 'origin'
DEBUG   -  Config value 'extra' = {}
DEBUG   -  Config value 'plugins' = {'mkdocstrings':
           <mkdocstrings.plugin.MkdocstringsPlugin object at
           0x7f7b2c147a70>}
DEBUG   -  Config value 'hooks' = {}
DEBUG   -  Config value 'watch' = []
DEBUG   -  Config value 'validation' = {'nav': {'omitted_files': 20,
           'not_found': 30, 'absolute_links': 20}, 'links':
           {'not_found': 30, 'absolute_links': 20,
           'unrecognized_links': 20}}
DEBUG   -  Running 1 `config` events
DEBUG   -  mkdocstrings: Adding extension to the list
DEBUG   -  mkdocstrings: Added a subdued autorefs instance
           <mkdocs_autorefs.plugin.AutorefsPlugin object at
           0x7f7b2c147c20>
DEBUG   -  mkdocs_autorefs: Adding AutorefsExtension to the list
INFO    -  Cleaning site directory
INFO    -  Building documentation to directory: /griffe-bug/site
DEBUG   -  Reading markdown pages.
DEBUG   -  Reading: index.md
DEBUG   -  Running 1 `page_markdown` events
DEBUG   -  Running 1 `page_content` events
DEBUG   -  Reading: bad.md
DEBUG   -  Running 1 `page_markdown` events
DEBUG   -  mkdocstrings: Matched '::: src.src.code_bad'
DEBUG   -  mkdocstrings: Using handler 'python'
DEBUG   -  mkdocstrings: Collecting data
DEBUG   -  griffe: Found src: loading
DEBUG   -  griffe: Loading path [PosixPath('src'),
           PosixPath('/griffe-bug/src')]
DEBUG   -  griffe: Skip src/src/code_bad.py, another module took
           precedence
DEBUG   -  griffe: Skip /griffe-bug/src/src/code_bad.py, another
           module took precedence
DEBUG   -  griffe: Loading path src/folder_1/code_good.py
DEBUG   -  griffe: Loading path /griffe-bug/src/folder_1/code_good.py
DEBUG   -  griffe: Trying to merge src/folder_1/code_good.py and
           /griffe-bug/src/folder_1/code_good.py
DEBUG   -  griffe: Iteration 1 finished, 0 aliases resolved, still 0
           to go
ERROR   -  mkdocstrings: src.src.code_bad could not be found
ERROR   -  Error reading page 'bad.md':
ERROR   -  Could not collect 'src.src.code_bad'

Aborted with a BuildError!
```

</details>

### Expected behavior

The `mkdocstrings`-process should handle a file/module that is placed independently of how the parent- and grand-parent-folder are named.
This used to work with griffe-versions before `0.39.0`.

### Environment information

```bash
python -m mkdocstrings.debug  # | xclip -selection clipboard
```

- __System__: Linux-6.5.0-14-generic-x86_64-with-glibc2.36
- __Python__: cpython 3.12.1
- __Environment variables__:
- __Installed packages__:
  - `mkdocstrings` v0.24.0

### Additional context

See https://github.com/IronTopi/griffe-bug for reproducing and additional description.

**Comment by pawamoy** [comment]
OK, it actually comes from the way I mark folders as "seen" when searching for Python modules within them. When I scan "src", I mark it as seen and recurse. It then finds another "src" folder and thinks it already saw it. A fix I can think of is to make path absolute (and maybe resolve them too), I'll see if there are any implications.

**Comment by pawamoy** [comment]
Fixed in 0.39.1
